samples/Exporters/Console/TestConsoleActivity.cs

```diff
@@ -0,0 +1,89 @@
+﻿// <copyright file="TestConsoleActivity.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Diagnostics;
+using OpenTelemetry.Exporter.Console;
+using OpenTelemetry.Trace.Configuration;
+
+namespace Samples
+{
+    internal class TestConsoleActivity
+    {
+        internal static object Run(ConsoleActivityOptions options)
+        {
+            // Enable OpenTelemetry for the source "MyCompany.MyProduct.MyWebServer"
+            // and use Console exporter
+            OpenTelemetrySdk.EnableOpenTelemetry(
+                (builder) => builder.AddActivitySource("MyCompany.MyProduct.MyWebServer")
+                .UseConsoleActivityExporter(opt => opt.DisplayAsJson = options.DisplayAsJson));
+
+            // The above line is required only in Applications
+            // which decide to use OT.
+
+            // Libraries would simply write the following lines of code to
+            // emit activities, which are the .NET representation of OT Spans.
+            var source = new ActivitySource("MyCompany.MyProduct.MyWebServer");
+
+            // The below commented out line shows more likely code in a real world webserver.
+            // using (var parent = source.StartActivity("HttpIn", ActivityKind.Server, HttpContext.Request.Headers["traceparent"] ))
+            using (var parent = source.StartActivity("HttpIn", ActivityKind.Server))
```

Comment:
I really do not like this API - it's very easy to missing a null check and introduce a non-obvious error. It also goes against the OpenTelemetry spec that a no-op tracer / span is returned in such a way no code needs to change.

Reply:
> it's very easy to missing a null check and introduce a non-obvious error. 

StartActivity is not different than any other API that can return a null value. we use the `?` for simplicity to simplify the code. Usually, users adopt new APIs will look at the doc and usage patterns. Also, usually, anyone writes code like that, is going to test it. simple case testing will catch this problem very easily.

> It also goes against the OpenTelemetry spec that a no-op tracer/span is returned in such a way no code needs to change.

There is no code that needs to change anyway. right? The code will be the same either there are listeners or not. please notice, Activity is a type that is used for many years. In our design, we wanted to serve both users already using Activity and OT scenarios. 

CC @noahfalk

Reply:
Hey @tarekgh - I didn't mean my comments to come across so negative, sorry. I'll try to be more constructive.

I guess my main issue is that it is significantly different in usage compared to the OpenTelemetry specification. The span & tracer interfaces are designed to be robust and easy to use, providing an efficient no-op design if configuration is incomplete.

The Activity API can return null and must be checked for manually every time a usr wishes to interact with it, which becomes error prone if using the shorthand null check. It also does not support non-string tags (see http status code) which deviates from the spec and will cause propagation issues between language implementations (all other languages allow non-strings).

Reply:
@MikeGoldsmith I appreciate your feedback, it is helpful. So, don't hesitate to send more :-) 

Yes, I understand there are some differences between the OT specs and the .NET APIs. We have been working hard to get them close and fill the gaps but at the same time, we are restricted by some other constraints. We had to come with a design that not confuses a lot of developers who currently using the Activity APIs and at the same allow the OT scenarios. Think about it as we are not strictly implementing the OT as it is but we are enabling all scenarios that OT can do. 

We'll continue to look at the feedback and try to figure out how we can help more with that. Thanks again for your feedback.

Reply:
> I didn't mean my comments to come across so negative

No worries at all @MikeGoldsmith, your initial feedback was straightforward and the objections were grounded. It was fine feedback : )

> I guess my main issue is that it is significantly different in usage compared to the OpenTelemetry specification. The span & tracer interfaces are designed to be robust and easy to use, providing an efficient no-op design if configuration is incomplete

I agree, it is different. From my perspective deciding to reuse Activity APIs and integrate into the platform rather than creating a competing API has a lot of advantages, but one disadvantage is that it creates additional constraints on the design. In this case OT's premise is that it is easy and cheap to create a no-op Span so that nobody has to do a null check. We considered a no-op Activity, but didn't like the options:
1. If the no-op Activity behaves like a normal Activity then it is stateful and we can't use a global singleton. This means we are paying for allocations and all the normal initialization. Some of the work unfortunately isn't that cheap.
2. If the no-op Activity doesn't behave like a normal Activity we can make it cheap, but now we've created some risk that pre-existing code assets had assumptions about how Activity works which are invalidated. For example maybe the existing code assumes that properties can be read back after they are set, or that the timestamp of a started activity isn't 0.

So we ended at option 3, which was let the user see the null result. Our hope is that compiler nullable reference analysis makes it obvious when you've done the wrong thing, or if you are using an older compiler then code snippets, docs, and testing. Certainly its possible that this will be more problematic than we estimated so getting this kind of feedback is important (thanks!). So far we haven't heard much concern raised around this, but our overall feedback volume is probably still low.


> It also does not support non-string tags

This one should probably be separated into its own issue. @cijothomas @reyang - have you guys experimented much with tags that would be non-string types in other languages? Certainly I see some potential for friction here.

Reply:
> This one should probably be separated into its own issue. @cijothomas @reyang - have you guys experimented much with tags that would be non-string types in other languages? Certainly I see some potential for friction here.

Yep. @cijothomas @tarekgh and I were chatting about this topic earlier today. We do see a good value of having non-string types.